### PR TITLE
Make Email Verification Code Length Customizable

### LINKED ISSUE #2247: Making the Email Verification Validation Code length Configurable in Serverpod Auth Module
### Discussed in https://github.com/serverpod/serverpod/discussions/2221

> 
> <div type='discussions-op-text'>
> 
> <sup>Originally posted by **adiltirur** May  8, 2024</sup>
> Hello everyone,
> 
> I've been working with the Serverpod auth module and I've noticed that the validationCode for email verification is currently set to 8 digits. For the purpose of enhancing user experience in mobile apps, I'm interested in making this length configurable, allowing for either 4 or 6 digit codes.
> 
> I'm wondering if there's already a way to configure the length of the validationCode that I might have missed. If not, could anyone provide guidance on how I might implement this feature?
> 
> Any suggestions or insights would be greatly appreciated. Thank you in advance for your help.</div>
> 

### Feature
Make it possible to configure the email verification validation code length to make it possible to improve mobile UI.

### PR DESCRIPTION
**Pull request by BenAuerDev**
This adds another field to the  `serverpod_auth_server` modules\` `AuthConfig`.
The length of the validation code  can now be configured by setting `validationCodeLength`  when declaring the `AuthConfig` in server.dart.

default is still 8.

When 0 or a negative value is passed an Exception is thrown. 

Tests were added to check:
- if valid lengths result in a code of that length being generated and Emails.createAccountRequest returns true
- if invalid lengths result in an exception being thrown in the constructor of AuthConfig 

auth-config-with-validationCodeLength

closes #2247

## Pre-launch Checklist

- [x] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [x] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [x] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [x] I added new tests to check the change I am making.
- [x] All existing and new tests are passing.
- [ ] Any breaking changes are documented below.

If you need help, consider asking for advice on the [discussion board](https://github.com/serverpod/serverpod/discussions).

## Breaking changes

No breaking changes

### CODE REVIEW

File: modules/serverpod_auth/serverpod_auth_server/lib/src/business/config.dart [hunk 1]
```diff
@@ -155,6 +159,7 @@ class AuthConfig {
     this.onUserUpdated,
     this.sendPasswordResetEmail,
     this.sendValidationEmail,
+    this.validationCodeLength = 8,
```
Comment (SandPod) [inline modules/serverpod_auth/serverpod_auth_server/lib/src/business/config.dart:162]: Discussion: Should we safeguard against setting the value to 0 or a negative value? Or what makes sense in that case?

**Review by SandPod — CHANGES_REQUESTED** [review]
One question around the implementation otherwise the fix looks good.

I would like you to add a test for this as well.
You can either add the test [here](https://github.com/serverpod/serverpod/blob/main/tests/serverpod_test_server/test_integration/authentication/email_auth_provider_test.dart) or add a file next to that specifically tests the AuthConfig changes (I prefer this).  Remember also to cover the default scenario in the test.

**Comment by SandPod** [comment]
We should also add some docs info that this is possible to configure.

**Comment by SandPod** [comment]
> We should also add some docs info that this is possible to configure.

If you are up for it, please feel free to document the full AuthConfig, then we could close this issue: https://github.com/serverpod/serverpod/issues/2245

A table might be a good way of documenting the configurable options.

**Comment by BenAuerDev** [comment]
> > We should also add some docs info that this is possible to configure.
> 
> If you are up for it, please feel free to document the full AuthConfig, then we could close this issue: #2245
> 
> A table might be a good way of documenting the configurable options.

Sure that is a great combo! 

Do you have any suggestions where this would fit best? I was thinking either authentication/setup maybe?

File: modules/serverpod_auth/serverpod_auth_server/lib/src/business/config.dart (hunk 1 above)
Comment (SandPod) [reply modules/serverpod_auth/serverpod_auth_server/lib/src/business/config.dart:162]: It might be good to add a suggestion for at least three digits with this motivation in the documentation. But I think, ultimately, the framework should not be in your way if you would like to use a single-digit code.

**Comment by SandPod** [comment]
> > > We should also add some docs info that this is possible to configure.
> > 
> > 
> > If you are up for it, please feel free to document the full AuthConfig, then we could close this issue: #2245
> > A table might be a good way of documenting the configurable options.
> 
> Sure that is a great combo!
> 
> Do you have any suggestions where this would fit best? I was thinking authentication/setup maybe?

That sounds good, maybe under the `Server setup` header?

File: modules/serverpod_auth/serverpod_auth_server/lib/src/business/config.dart (hunk 1 above)
Comment (BenAuerDev) [reply modules/serverpod_auth/serverpod_auth_server/lib/src/business/config.dart:162]: True :) But this sounds like an elegant solution to me. 

Safeguarding against `0` or `negative` 

Recommending / informing in the docs

**Comment by BenAuerDev** [comment]
let me know if/where there is room for improvement :)

File: tests/serverpod_test_server/test_integration/authentication/auth_config_changes_test.dart [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+void main() async {
+  var userName = 'test';
+  var email = 'test@serverpod.dev';
+  var password = 'password';
+  var session = await IntegrationTestServer().session();
+
+  var codeLengthsToTest = [-16, -8, -4, -1, 0, 1, 4, 16];
+
+  test(
+      'Not setting validationCodeLength will generate a code with default length 8',
```
Comment (SandPod) [inline tests/serverpod_test_server/test_integration/authentication/auth_config_changes_test.dart:14]: Suggestion: Following the `Given, When, Then` pattern a suggestion for this description could be "Given no validation code length configuration when creating an account then validation code has the default length of 8.".

File: tests/serverpod_test_server/test_integration/authentication/auth_config_changes_test.dart [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+      isTrue,
+      reason: 'Generated validation code is 8 characters long, which is valid',
+    );
+
+    assert(generatedValidationCode?.length == 8);
+  });
+
+  for (var length in codeLengthsToTest) {
+    final willPass = length >= 1;
+    test(
```
Comment (SandPod) [inline tests/serverpod_test_server/test_integration/authentication/auth_config_changes_test.dart:40]: Suggestion: Please break these out into individual explicit tests.

In tests it is fine to have code duplication, it is all about making sure the developer reading the test understand what is being tested and why.

I also think that it should be enough with three tests, one using a negative value, one using zero and one using a positive value :)

File: modules/serverpod_auth/serverpod_auth_server/lib/src/business/email_auth.dart [hunk 4]
```diff
@@ -578,8 +578,16 @@ class Emails {
   }
 
   static String _generateVerificationCode() {
+    var lenValidationCode = AuthConfig.current.validationCodeLength;
+
+    if (lenValidationCode <= 0) {
```
Comment (SandPod) [inline modules/serverpod_auth/serverpod_auth_server/lib/src/business/email_auth.dart:583]: Suggestion: Move the validation into the constructor of the AuthConfig.

This won't throw until the server tries to issue a validation code. If we move it into the AuthConfig we will throw an error directly when the server is starting which should help the developer identify the configuration issue directly.

File: tests/serverpod_test_server/test_integration/authentication/auth_config_changes_test.dart [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+      String? generatedValidationCode;
+      AuthConfig.set(
+        AuthConfig(
+          validationCodeLength: length,
+          sendValidationEmail: (session, email, validationCode) async {
+            generatedValidationCode = validationCode;
+            return true;
+          },
+        ),
+      );
```
Comment (SandPod) [inline tests/serverpod_test_server/test_integration/authentication/auth_config_changes_test.dart:54]: Discussion: If we move the throw into the constructor we can expect that the AuthConfig constructor throws directly instead of having to rely on the `set` method and call `createAccountRequest`. 

You can see some example code for expecting exceptions here if you are not familiar with it from before: https://github.com/serverpod/serverpod/blob/a29ecdb8221020afc433f5b94612620d08607d89/packages/serverpod/test/database/table_relation_test.dart#L9-L19

**Review by SandPod — CHANGES_REQUESTED** [review]
Some changes are needed for the test to make them as simple as possible to understand and it would be great to have the throw in the constructor instead.

File: modules/serverpod_auth/serverpod_auth_server/lib/src/business/email_auth.dart (hunk 4 above)
Comment (BenAuerDev) [reply modules/serverpod_auth/serverpod_auth_server/lib/src/business/email_auth.dart:583]: Agree! This is a much better approach, as now (when thrown in constructor of AuthConfig) the Exception when the server tries to start.

File: tests/serverpod_test_server/test_integration/authentication/auth_config_changes_test.dart (hunk 5 above)
Comment (BenAuerDev) [reply tests/serverpod_test_server/test_integration/authentication/auth_config_changes_test.dart:54]: Yes this is a much better approach! Thanks :)  

Just to make sure: For the test cases that are not supposed to throw an Exception can I just check for:

`expect(() => AuthConfig(validationCodeLength: 4), 4);`

Or should i still call `createAccountRequest` there?

File: modules/serverpod_auth/serverpod_auth_server/lib/src/business/config.dart [hunk 6]
```diff
@@ -155,12 +159,19 @@ class AuthConfig {
...
         'config/firebase_service_account_key.json',
     this.maxPasswordLength = 128,
     this.minPasswordLength = 8,
     this.allowUnsecureRandom = false,
-  });
+  }) {
+    if (validationCodeLength < 1) {
+      throw Exception(
+        'Invalid validation code length: $validationCodeLength. Length must be at least 1.',
+      );
```
[the remark this answers is not in the document]
Comment (BenAuerDev) [reply modules/serverpod_auth/serverpod_auth_server/lib/src/business/config.dart:174]: I honestly had to look it up, but `ArgumentError.value()` seems to be a better choice, as it is more specific. 

And the console output is nicer :D

**ArgumentError.value():**
Unhandled exception:
Invalid argument (validationCodeLength): must be at least 1: 0
#0      new AuthConfig (package:serverpod_auth_server/src/business/config.dart:172:7)



**Exception:**
Unhandled exception:
Exception: Invalid validation code length: 0. Length must be at least 1.
#0      new AuthConfig (package:serverpod_auth_server/src/business/config.dart:172:7)

**Review by SandPod — APPROVED** [review]
The changes look good now!

Awesome contribution 🚀

**Comment by BenAuerDev** [comment]
> The changes look good now!
> 
> Awesome contribution 🚀

Thanks I learned a lot :rocket:

**Comment by SandPod** [comment]
But please also add a PR to our documentation repo for this change and if possible document the rest of the AuthConfig :)

**Comment by BenAuerDev** [comment]
Already on it :) 

I went for a table as you suggested. :) Let me know if you have any input :) 

configure-authentication
